test/test.ml

```diff
@@ -11,205 +11,128 @@
   for _ = 1 to nb_iter do
-    let out1 = kCAS c1 in
-    let out2 = kCAS c2 in
-    if out1 <> true || out2 <> true then begin
-(*      print_endline (sprintf "TH%d  APPEL N°%d ECHEC OUT1 = %b    OUT2 = %b!!!!!!!!!" (Domain.self ()) i out1 out2);*)
-      th1_success := false
-    end
+    let out1 = Kcas.kCAS c1 in
+    let out2 = Kcas.kCAS c2 in
+    assert out1;
+    assert out2
   done;
-  ignore @@ kCAS c1;
-  ()
-;;
+  ignore @@ Kcas.kCAS c1
 
 let thread2 (a1, a2) =
-  let c1 = [mk_cas a1 v_y v_x ; mk_cas a2 v_x v_y] in
-  let c2 = [mk_cas a1 v_x v_y ; mk_cas a2 v_y v_x] in
+  let c1 = [ Kcas.mk_cas a1 1 0; Kcas.mk_cas a2 0 1 ] in
+  let c2 = [ Kcas.mk_cas a1 0 1; Kcas.mk_cas a2 1 0 ] in
   for _ = 1 to nb_iter do
-    let out1 = kCAS c1 in
-    let out2 = kCAS c2 in
-    if out1 <> false || out2 <> false then begin
-(*      print_endline (sprintf "TH%d  APPEL N°%d ECHEC OUT1 = %b    OUT2 = %b!!!!!!!!!" (Domain.self ()) i out1 out2);*)
-      th2_success := false
-    end
+    let out1 = Kcas.kCAS c1 in
+    let out2 = Kcas.kCAS c2 in
+    assert (not out1);
+    assert (not out2)
   done
-;;
 
 let thread3 (a1, a2) =
-  let c1 = [mk_cas a1 v_x v_y ; mk_cas a2 v_y v_x] in
-  let c2 = [mk_cas a1 v_y v_x ; mk_cas a2 v_x v_y] in
+  let c1 = [ Kcas.mk_cas a1 0 1; Kcas.mk_cas a2 1 0 ] in
+  let c2 = [ Kcas.mk_cas a1 1 0; Kcas.mk_cas a2 0 1 ] in
   for _ = 1 to nb_iter do
-    let out1 = kCAS c1 in
-    let out2 = kCAS c2 in
-    if out1 <> false || out2 <> false then begin
-(*      print_endline (sprintf "TH%d  APPEL N°%d ECHEC OUT1 = %b    OUT2 = %b!!!!!!!!!" (Domain.self ()) i out1 out2);*)
-      th3_success := false
-    end
+    let out1 = Kcas.kCAS c1 in
+    let out2 = Kcas.kCAS c2 in
+    assert (not out1);
+    assert (not out2)
   done
-;;
 
 let thread4 (a1, a2) =
   for i = 0 to nb_iter do
-    let c = [mk_cas a1 i (i+1) ; mk_cas a2 i (i+1)] in
-    let out = kCAS c in
-    if out <> true then
-      th4_success := false
+    let c = [ Kcas.mk_cas a1 i (i + 1); Kcas.mk_cas a2 i (i + 1) ] in
+    assert (Kcas.kCAS c)
   done
-;;
 
 let thread5 (a1, a2) =
   for _ = 0 to nb_iter do
-    let a = get a1 in
-    let b = get a2 in
-    if a > b then
-      th5_success := false
+    let a = Kcas.get a1 in
+    let b = Kcas.get a2 in
+    assert (a <= b)
   done
-;;
 
 let test_casn () =
-  let a1 = ref v_x in
-  let a2 = ref v_x in
-
-  Domain.spawn (fun () -> thread1 (a1, a2)) |> ignore;
-  Domain.spawn (fun () -> thread2 (a1, a2)) |> ignore;
-  Domain.spawn (fun () -> thread3 (a1, a2)) |> ignore;
-
-  Unix.sleep wait_time;
-  print_endline (sprintf "a1 = %d et a2 = %d" (get a1) (get a2));
-  !th1_success && !th2_success && !th3_success
-;;
+  let a1 = Kcas.ref 0 in
+  let a2 = Kcas.ref 0 in
+
+  let domains =
+    [
+      (fun () -> thread1 (a1, a2));
+      (fun () -> thread2 (a1, a2));
+      (fun () -> thread3 (a1, a2));
+    ]
+  in
+  List.map Domain.spawn domains |> List.iter Domain.join
```

Comment:
Again, I realise this has not fundamentally changed in this PR, but this test has a race condition.  It is (theoretically) possible that each domain gets spawned and completes before other domains.

Reply:
There are many ways to fix this.

What I might suggest is to make it so that `thread2` and `thread3` are changed to wait for a condition A before starting the interference loop.  Instead of performing a fixed amount of iterations, `thread2` and `thread3` would run their loops until another condition B becomes true.  `thread1` then first makes the condition A true and then runs the loop after which it makes condition B true.  Additionally `thread2` and `thread3` could increment counters that `thread1` checks periodically, perhaps two nested loops in `thread1`, for having been actually incremented during the inner loop to check that `thread2` and `thread3` are actually running during the inner loop in `thread1`.